### ISSUE TITLE
Don't allow use of && and || on vectors

### ISSUE THREAD

**Issue by mungojam** [hadley/strict#32]
This one catches me out quite a lot as I've come from another language where && and || are almost exclusively used (since they just mean short-circuit and/or).

```r
x <- c(1,2,3,4,5)
x > 1 && x < 4
#[1] FALSE
```

```r
x <- c(1,2,3,4,5)
x >= 1 & x < 6
[1] TRUE TRUE TRUE TRUE TRUE
```

I realise now that the first one is meant for testing if all elements match but that feels like magic. But for that, I'd happily be forced to use `all()` which incidentally works whether I use & or && (but again I'd be happy to be forced to use &):

```r
x <- c(1,2,3,4,5)
all(x > 1 && x < 4)
#[1] FALSE

all(x > 1 & x < 4)
#[1] FALSE

all(x >= 1 && x < 6)
#[1] TRUE

all(x >= 1 & x < 6)
#[1] TRUE
```

**Comment by WerthPADOH** [comment]
> I realise now that the first one is meant for testing if all elements match

It's actually just returning the first element of a logical vector.

```
> x > 0 && x < 4
[1] TRUE
```

Still, I agree it makes sense to cover this.  Same with `if (vector)`, which is nice enough in base R to warn when a vector's given.

**Comment by mungojam** [comment]
You're right thanks, convinced myself otherwise with my inconclusive test.

Yeah, covering if(vector) too would be cool, I've hit that too before a few
times.
